crypto/ec/ecdsa_ossl.c

```diff
@@ -153,30 +153,33 @@ static int ecdsa_sign_setup(EC_KEY *eckey, BN_CTX *ctx_in,
     }
     while (BN_is_zero(r));
 
-    /* compute the inverse of k */
-    if (EC_GROUP_get_mont_data(group) != NULL) {
-        /*
-         * We want inverse in constant time, therefore we utilize the fact
-         * order must be prime and use Fermat's Little Theorem instead.
-         */
-        if (!BN_set_word(X, 2)) {
-            ECerr(EC_F_ECDSA_SIGN_SETUP, ERR_R_BN_LIB);
-            goto err;
-        }
-        if (!BN_mod_sub(X, order, X, order, ctx)) {
-            ECerr(EC_F_ECDSA_SIGN_SETUP, ERR_R_BN_LIB);
-            goto err;
-        }
-        BN_set_flags(X, BN_FLG_CONSTTIME);
-        if (!BN_mod_exp_mont_consttime
-            (k, k, X, order, ctx, EC_GROUP_get_mont_data(group))) {
-            ECerr(EC_F_ECDSA_SIGN_SETUP, ERR_R_BN_LIB);
-            goto err;
-        }
-    } else {
-        if (!BN_mod_inverse(k, k, order, ctx)) {
-            ECerr(EC_F_ECDSA_SIGN_SETUP, ERR_R_BN_LIB);
-            goto err;
+    /* Check if optimized inverse is implemented */
+    if (EC_GROUP_do_inverse_ord(group, k, k, ctx) == 0) {
```

Comment:
Sorry quick question. I assume the new method called from EC_GROUP_do_inverse_ord is in constant time where necessary as the comment at line 161 seems to suggest a constant time implementation is needed when doing the inverse?
If it does need to be in constant time is it worth adding a comment somewhere for the new method that it should be in constant time as it is a method that could be overidden for an engine implementation for instance?
Apologies if I am not understanding correctly!

Reply:
In this specific context const-time-ness is not concern, but in general case it might be appropriate to pass it down to implementation, yes. I mean the fact if method is called from sign or verify procedures. I'd suggest to address it in separate merge request....

Reply:
@dot-asm I'm not sure what you mean. In this context it is essential that the inversion be done in constant time. It even says so below: " "We want inverse in constant time"

Reply:
I meant that the specific method implementation, ecp_nistz256_inv_mod_ord, not EC_GROUP_do_inverse_ord, doesn't differentiate sign from verify operations, and treats even verify operations as if they were sign ones. [And sign operations are constant time.] One can also put it like this, *when* we add extra flag, this specific method implementation will ignore it.